pkg/erofs/erofs.go

```diff
@@ -690,87 +724,155 @@ func (i *Inode) blocks() uint64 {
+	if err != nil {
+		return nil, err
+	}
+	if lastDirent {
+		// Optional padding may exist at the end of a block.
+		n := bytes.IndexByte(name, 0)
+		if n == 0 {
+			log.Warningf("Corrupted dirent at inode (nid=%v)", i.Nid())
+			return nil, linuxerr.EUCLEAN
+		}
+		if n != -1 {
+			name = name[:n]
+		}
+	}
+	return name, nil
+}
+
+// getDirent0 returns a pointer to the first dirent in the given block of this inode.
+func (i *Inode) getDirent0(block blockData) (*Dirent, error) {
+	d0, err := i.image.direntAt(block.base)
+	if err != nil {
+		return nil, err
+	}
+	// Apart from the offset of the first filename, nameOff0 also indicates
+	// the total number of dirents in this block.
+	if d0.NameOff < DirentSize || uint32(d0.NameOff) >= block.size {
+		log.Warningf("Invalid nameOff0 %v at inode (nid=%v)", d0.NameOff, i.Nid())
+		return nil, linuxerr.EUCLEAN
+	}
+	return d0, nil
+}
+
+// Lookup looks up a child by the name. The child inode number will be returned on success.
+func (i *Inode) Lookup(name string) (uint64, error) {
 	if !i.IsDir() {
-		return linuxerr.ENOTDIR
+		return 0, linuxerr.ENOTDIR
 	}
 
-	blocks := i.blocks()
-	blockSize := i.image.BlockSize()
+	// Currently (Go 1.21), there is no safe and efficient way to do three-way
+	// string comparisons, so let's convert the string to a byte slice first.
+	nameBytes := gohacks.ImmutableBytesFromString(name)
 
-	start := i.dataOff
-	if blocks == 1 && i.idataOff != 0 {
-		start = i.idataOff
+	// In EROFS, all directory entries are _strictly_ recorded in alphabetical
+	// order. The lookup is done by directly performing binary search on the
+	// disk data similar to what Linux does in fs/erofs/namei.c:erofs_namei().
+	type blockInfo struct {
+		block      blockData
+		numDirents uint16
 	}
+	var target blockInfo
 
-	// Iterate all the blocks which contain dirents.
-	for blocks > 0 {
-		// Get the first dirent in the current block.
-		direntOff := start
+	// Find the block that may contain the target dirent first.
+	left, right := int64(0), int64(i.blocks)-1
+	for left <= right {
+		// Cast to uint64 to avoid overflow.
+		mid := uint64(left+right) >> 1
+		block := i.getBlockDataInfo(mid)
+		d0, err := i.getDirent0(block)
+		if err != nil {
+			return 0, err
+		}
+		numDirents := d0.NameOff / DirentSize
+		name0, err := i.getDirentName(d0, block, numDirents == 1)
+		if err != nil {
+			return 0, err
+		}
+		switch bytes.Compare(nameBytes, name0) {
+		case 0:
+			// Found the target dirent.
+			return d0.Nid(), nil
+		case 1:
+			// name > name0, this block may contain the target dirent.
+			target = blockInfo{block, numDirents}
+			left = int64(mid) + 1
+		case -1:
+			// name < name0, this is not the block we're looking for.
+			right = int64(mid) - 1
+		}
+	}
+
+	if !target.block.valid() {
+		// The target block was not found.
+		return 0, linuxerr.ENOENT
+	}
+
+	// Find the target dirent in the target block.
+	left, right = 1, int64(target.numDirents)-1
```

Comment:
In this case, `left` and `right` can be just uint16s? It will avoid a bunch of casts. Maybe name the two lefts and rights differently: `b{Left/Right}` for block left right, and `d{Left/Right}` for dirent left right.

Here `mid := uint16((uint32(left)+uint32(right)) >> 1)`

Also add comment that we don't check 0th dirent as it was checked during block binary search.

Reply:
Good idea. Will name the two lefts and rights differently and add the comment you suggested. Thanks!